File: DAY_01/Devraj/surrounded_region.cpp
class Solution {
public:
    int X[4]={1,-1,0,0};
    int Y[4]={0,0,1,-1};
    void dfs(vector<vector<char>>& board,int x,int y,int n,int m)
    {    if(x<0 || y<0 || x>= n || y>=m || board[x][y]=='X' || board[x][y]=='V')
           return;
        
        board[x][y]='V';
        for(int i=0;i<4;i++)
        {   int newx=x+X[i];
            int newy=y+Y[i];
            dfs(board,newx,newy,n,m);
           
        }
        
    }
    void solve(vector<vector<char>>& board) {
        
        int n=board.size();int m=board[0].size();
        for(int i=0;i<n;i++)
        {   if(board[i][0]=='O')           //DFS for both left and right boundary
              dfs(board,i,0,n,m);
            if(board[i][m-1]=='O')
              dfs(board,i,m-1,n,m);
                 
        }
         for(int i=0;i<m;i++)
        {   if(board[0][i]=='O')
              dfs(board,0,i,n,m);                         //DFS for both upper and lower boundary
            if(board[n-1][i]=='O')
              dfs(board,n-1,i,n,m);
            
        }
        // for(int i=0;i<n;i++)
        // {
        //     for(int j=0;j<m;j++)
        //         cout<<board[i][j]<<" ";
        //     cout<<"\n";
        // }
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                  if(board[i][j]=='V')                  // If visited in DfS,Then convert it to O again
                      board[i][j]='O';
                  else if(board[i][j]=='O')
                        board[i][j]='X';                   // If 0 Then convert it to X
                
                
                
            }
        }
        
       
    }
};

File: DAY_01/Peeyush_Varyani/Surrounded_Region.java
public class Solution {
    public void solve(char[][] board) {
        int rows = board.length;
        if(rows == 0) return;
        int cols = board[0].length;
        
        for(int i = 0; i < cols; i++) {
            // check first row's O
            if(board[0][i] == 'O') {
                // change it to other symbol
                board[0][i] = '#';
                dfs(board, 0, i);
            }
            // check the last row
            if(board[rows - 1][i] == 'O') {
                board[rows - 1][i] = '#';
                dfs(board, rows - 1, i);
            }

        }
        
        for(int i = 0; i < rows; i++) {
            // check first col
            if(board[i][0] == 'O') {
                board[i][0] = '#';
                dfs(board, i, 0);
            }
            // check last col
            if(board[i][cols - 1] == 'O') {
                board[i][cols - 1] = '#';
                dfs(board, i, cols - 1);
            }
        }
        
        // change O to X
        changeTo(board, 'O', 'X');
        
        // change # to O
        changeTo(board, '#', 'O');
        return;
    }
    
    public void dfs(char[][] board, int row, int col) {
        // check up
        if(row > 0) {
            if(board[row - 1][col] == 'O') {
                board[row - 1][col] = '#';
                dfs(board, row - 1, col);
            }
        }
        // check left
        if(col > 0) {
            if(board[row][col - 1] == 'O') {
                board[row][col - 1] = '#';
                dfs(board, row, col - 1);
            }
        }
        // check right
        if(row < board.length - 1) {
            if(board[row + 1][col] == 'O') {
                board[row+1][col] = '#';
                dfs(board, row+1, col);
            }
        }
        // check down
        if(col < board[0].length - 1) {
            if(board[row][col+1] == 'O'){
                board[row][col+1] = '#';
                dfs(board, row, col+1);
            }
        }
        return;
    }
    
    public void changeTo(char[][] board, char from, char to) {
        for(int i = 0; i < board.length; i++) {
            for(int j = 0; j < board[0].length; j++) {
                if(board[i][j] == from) {
                    board[i][j] = to;
                }
            }
        }
        return;
    }
}
